### PR TITLE
Respond to push webhook event

### PR DESCRIPTION
**Pull request by ErikSchierboom**

### CODE REVIEW

File: app/controllers/webhooks/base_controller.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# This is the base controller of Exercism's webhooks.
+#
+# It is intended to only be called from GitHub webhooks
+# and should not be called by any other source
+
+module Webhooks
+  class BaseController < ApplicationController
+    skip_before_action :verify_authenticity_token
+    skip_before_action :authenticate_user!
```
Comment (ErikSchierboom) [inline app/controllers/webhooks/base_controller.rb:9]: Is this okay? We don't have need of a user or or CSRF tokens, as we do the signature validation. This is also how the SPI controller does it.

File: app/controllers/webhooks/base_controller.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    private
+    def verify_github_webhook
+      return if signature_valid?
+
+      render json: {
+        error: {
+          type: :invalid_webhook_signature,
+          message: "The auth token provided is invalid. Delivery: #{github_delivery}"
+        }
+      }, status: :internal_server_error
```
Comment (ErikSchierboom) [inline app/controllers/webhooks/base_controller.rb]: The API controllers use I8N for the error messages. I could do that here, but it didn't know how to add in the delivery argument (which I've copied from the v2 code).

File: app/controllers/webhooks/base_controller.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      Exercism.secrets.github_webhooks_secret
+    end
+
+    def github_signature
+      request.headers['HTTP_X_HUB_SIGNATURE_256']
+    end
+
+    def github_delivery
+      request.headers['HTTP_X_GITHUB_DELIVERY']
+    end
```
Comment (ErikSchierboom) [inline app/controllers/webhooks/base_controller.rb:47]: These headers (and the signature calculating code) have been updated to the code specified by GitHub: https://developer.github.com/webhooks/securing/#validating-payloads-from-github

File: app/controllers/webhooks/base_controller.rb (hunk 1 above)
Comment (iHiD) [reply app/controllers/webhooks/base_controller.rb:9]: Yep. This seems correct! :)

File: app/controllers/webhooks/base_controller.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+#
+# It is intended to only be called from GitHub webhooks
+# and should not be called by any other source
+
+module Webhooks
+  class BaseController < ApplicationController
+    skip_before_action :verify_authenticity_token
+    skip_before_action :authenticate_user!
+
+    before_action :verify_github_webhook
```
Comment (iHiD) [inline app/controllers/webhooks/base_controller.rb]: Throw a `!` on the end of this as it's an explosive action.

Comment (ErikSchierboom) [reply app/controllers/webhooks/base_controller.rb]: Bang!

File: app/controllers/webhooks/base_controller.rb (hunk 2 above)
Comment (iHiD) [reply app/controllers/webhooks/base_controller.rb]: I don't think we need i18n here as it's not public facing.

File: app/controllers/webhooks/repo_updates_controller.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    def track_slug
+      params[:repo][:name]
+    end
+
+    def pushed_to_master?
+      params[:ref] == MASTER_REF
+    end
+
+    MASTER_REF = "refs/heads/master".freeze
+    private_constant :MASTER_REF
```
Comment (iHiD) [inline app/controllers/webhooks/repo_updates_controller.rb]: So none of this should get leaked into the controller itself. The controller is responsible for taking the request and handing it to to something to process, and then rendering the response. It shouldn't have any logic about the handling itself. 

Defining these methods inside the controller also leak them into every other controller action that might live in this at some time. As a general (pretty-hard) rule, controllers should only contain methods for actions, and methods for shared code between those actions (e.g. auth or retrieval of record across things).

So this should be moved either inline in the controller method, or should be moved into `app/commands/webhooks/process_repo_update`. The second would be my preference in this scenario.

File: test/controllers/webhooks/base_test_case.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+module Webhooks
+  class BaseTestCase < ActionDispatch::IntegrationTest
+    def headers(payload)
+      delivery = SecureRandom.uuid
+      signature = 'sha256=' + OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new('sha256'),
+        Exercism.secrets.github_webhooks_secret, payload.to_json)
+
+      @headers = {
+        'HTTP_X_GITHUB_DELIVERY' => delivery,
```
Comment (iHiD) [inline test/controllers/webhooks/base_test_case.rb]: You can inline the `SecureRandom.uuid` here as it's only used here.

File: test/controllers/webhooks/repo_updates_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+      ref: 'refs/heads/master',
+      repo: { name: 'csharp' }
+    }
+
+    invalid_headers = headers(payload)
+    invalid_headers['HTTP_X_HUB_SIGNATURE_256'] = "invalid_signature"
+
+    post webhooks_repo_updates_path, headers: invalid_headers, as: :json, params: payload
+    assert_response 500
+  end
```
Comment (iHiD) [inline test/controllers/webhooks/repo_updates_controller_test.rb:15]: This should probably be 403 not a 500. Reserve 500s for actual unexpected errors. That allows us to then capture those via Bugsnag and fix them all.

File: app/controllers/webhooks/repo_updates_controller.rb (hunk 5 above)
Comment (iHiD) [reply app/controllers/webhooks/repo_updates_controller.rb]: (The command should probably take the `params[:repo]` and `params[:ref]` as params)

File: app/controllers/webhooks/base_controller.rb (hunk 3 above)
Comment (iHiD) [reply app/controllers/webhooks/base_controller.rb:47]: Nice.

**Pull request merged**